### ISSUE TITLE
Move initial configuration to `Configure` in .NET Core

### ISSUE THREAD

**Issue by rkapka** [AutoMapper/AutoMapper#2748]
I don't believe it is currently possible to correctly register dependency injection with AutoMapper in .NET Core. Consider the following code inside `ConfigureServices`:
```
services.AddAutoMapper(cfg =>
{
     cfg.ConstructServicesUsing(services.BuildServiceProvider().GetService);
});
```
I have a value resolver that depends on Entity Framework's database context. The call to `services.BuildServiceProvider()` returns a new instance of service provider that has its own collection of services. Here is the relevant code from `ServiceCollectionContainerBuilderExtensions`:

```return new ServiceProvider((IEnumerable<ServiceDescriptor>) services, options);```

I don't think there is any other way to get a service provider inside `Configure`. This means that AutoMapper will inject a different instance of database context than the one used by my domain services. This causes issues when both contexts use the same entity. I can probably hack my way around this somehow, but the same issue applies to all injected services. If you set up a service as a singleton, then you end up with two "singletons" (I have verified this using a static field incremented in the constructor).

I think this can be resolved by adding AutoMapper middleware in `ConfigureServices` and configuring it in `Configure`. `Configure` has access to the default, and already instantiated, service provider. That way you could pass this provider to `ConstructServicesUsing`.

To sum it up, the code could look something like this:
```
public void ConfigureServices(IServiceCollection services)
{
    services.AddAutoMapper();
}

public void Configure(IApplicationBuilder app, IHostingEnvironment env)
{
    app.UseAutoMapper(cfg =>
    {
        cfg.ConstructServicesUsing(app.ApplicationServices.GetService);
    });
}

**Comment by lbargaoanu** [comment]
That's a different repository.
